Improvements to avfilter interface and color handling

jstebbins opened the pull request:
Simplifies the internal API to avfilter.
Reworks how avfilter based hb filters are initialized.
Improves logging of avfilter based hb filters.
Adds colorspace conversion filter (currently unused and untested)
Propagate colorspace information everywhere that it might be needed.

- [x] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [x] Ubuntu Linux

jstebbins committed:
avfilter: simplify internal API to avfilter

Use this interface in decavcodec.
Remove cropscale.c and use avfilter for crop and scale.
Improve log output of filters that are aliases for avfilter.

jstebbins committed:
libhb: propagate color matrix info through the pipeline

Propagates pix_fmt, range, primaries, transfer, and matrix everywhere.

Everything that passes or creates video frames tags the frames with
their color matrix info.

All filters know the expected color matrix info of input frames.

jstebbins commented:
I forgot to mention, eliminates cropscale.c by using avfilter.

jstebbins committed:
Add missing yuv444p12 colorspace opton

bradleysepos commented on libhb/hbavfilter.h:
```diff
@@ -0,0 +1,28 @@
+/* hbffmpeg.h
```
hbavfilter.h

bradleysepos commented:
Initial read looks good. Very promising!

jstebbins committed:
libhb: change preset key to VideoColorMatrixCodeOverride

Reserve VideoColorMatrixCode for doing color conversion.
VideoColorMatrixCodeOverride sets color flags without transforming video
frame colors.

jstebbins committed:
avfilter: fix header file comment

jstebbins committed:
avfilter: split individual filters to separate files

Splits the initialization of pad, cropscale, deinterlace, rotate, and
colorspace into their own files.  Also splits internal avfilter API into
it's own file.

jstebbins committed:
hbavfilter: hide internals of hb_avfilter_graph_s

jstebbins committed:
hbavfilter: fix filename in header comment

Nomis101 commented:
I've tested this on macOS 10.13 with some different video files. I found there is an issue with movies which are from UHD Blu-Rays (4k, HDR10). If I convert one of those into 10-bit HEVC, the resulting video is just green (I know, it doesn't make sense to convert 10-bit HEVC into 10-bit HEVC, but this was just a test).

Source file:
Stream #0:0(eng): Video: hevc (Main 10) (hvc1 / 0x31637668), yuv420p10le(tv, bt2020nc/bt2020/smpte2084), 3840x2160 [SAR 1:1 DAR 16:9], 49979 kb/s, 23.98 fps, 23.98 tbr, 16k tbn, 23.98 tbc (default)

Output file:
Stream #0:0(und): Video: hevc (Main 10) (hvc1 / 0x31637668), yuv420p10le(tv, bt2020nc/bt2020/smpte2084), 3840x1600 [SAR 1:1 DAR 12:5], 22780 kb/s, 23.98 fps, 23.98 tbr, 90k tbn, 23.98 tbc (default)

No issue with current master.

jstebbins commented:
Any chance you could clip a few seconds of the source and supply a sample?

I can't think how the output would be different than current master since in theory nothing has changed aside from passing the color information through the pipeline.

jstebbins commented:
I found some HDR10 samples online.  I'll try those once they are downloaded.

Samples:
https://doc-10-1s-docs.googleusercontent.com/docs/securesc/l4e7c6jrfkq7ipqflgi4o0bet9io960v/f2tpttl9mlibvrukb030abc0mgk2iunc/1554307200000/05996297375205892982/18308349273564643788/0Bxj6TUyM3NwjT0hEVW1CZ0xKaTg?e=download

https://doc-10-1s-docs.googleusercontent.com/docs/securesc/l4e7c6jrfkq7ipqflgi4o0bet9io960v/n9slql084t0lftavcq8hcsb4hmvm2idf/1554307200000/05996297375205892982/18308349273564643788/0Bxj6TUyM3NwjTmFBT2g2bV9zTlU?e=download

https://doc-10-1s-docs.googleusercontent.com/docs/securesc/l4e7c6jrfkq7ipqflgi4o0bet9io960v/fa4h42leo9rmi00i3gi488lmt801hfp6/1554307200000/05996297375205892982/18308349273564643788/0Bxj6TUyM3NwjQTRjTk9Vck1TT00?e=download&nonce=1ueld9etl988u&user=18308349273564643788&hash=kgkiif9gviamijo8anlp0ag009o42h24

And some HDR10 calibration patterns
https://www.avsforum.com/forum/139-display-calibration/2943380-hdr10-test-patterns-set.html

jstebbins commented:
I've reproduced the issue, thanks for doing some testing and pointing this out.

jstebbins committed:
cropscale: force output pixfmt to YUV420P

It appears some color prim/tran/matrix settings result in high bit
depth output of the scale filter.

jstebbins commented:
Fixed

bradleysepos commented:
You've invented a new planar format! :smile_cat:

Nomis101 commented:
> Fixed

Yes, seems fixed. :-)
